service/controller/v13/resource/node/create.go

```diff
@@ -0,0 +1,125 @@
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c := k8srestconfig.Config{
+			Logger: r.logger,
+
+			Address:   key.ClusterAPIEndpoint(customObject),
+			InCluster: false,
+			TLS: k8srestconfig.TLSClientConfig{
+				CAData:  certs.APIServer.CA,
+				CrtData: certs.APIServer.Crt,
+				KeyData: certs.APIServer.Key,
+			},
+		}
+		restConfig, err := k8srestconfig.New(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		k8sClient, err = kubernetes.NewForConfig(restConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "created K8s client for the guest cluster")
+	}
+
+	// We need to fetch the nodes being registered within the guest cluster's
+	// Kubernetes API. The list of nodes is used below to sort out which ones have
+	// to be deleted if there does no associated host cluster pod exist.
+	var nodes []corev1.Node
+	{
+		list, err := k8sClient.CoreV1().Nodes().List(metav1.ListOptions{})
+		if IsGuestAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "guest cluster is not available")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+		nodes = list.Items
+	}
+
+	// Fetch the list of pods running on the host cluster. These pods serve VMs
+	// which in turn run the guest cluster nodes. We use the pods to compare them
+	// against the guest cluster nodes below.
+	var pods []corev1.Pod
+	{
+		n := key.ClusterID(customObject)
+		list, err := r.k8sClient.CoreV1().Pods(n).List(metav1.ListOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		pods = list.Items
+	}
+
+	// Iterate through all nodes and compare them against the pods of the host
+	// cluster. Nodes being in a Ready state are fine. Nodes that belong to host
+	// cluster pods are also ok. If a guest cluster node does not have an
+	// associated host cluster pod, we delete it from the guest cluster's
+	// Kubernetes API.
+	for _, n := range nodes {
+		if node.IsNodeReady(&n) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not deleting node '%s' because it is in state 'Ready'", n.GetName()))
+			continue
+		}
+
+		if doesNodeExistAsPod(pods, n) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not deleting node '%s' because its host cluster pod does exist", n.GetName()))
+			continue
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting node '%s' in the guest cluster's Kubernetes API", n.GetName()))
+
+		err = k8sClient.CoreV1().Nodes().Delete(n.GetName(), nil)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted node '%s' in the guest cluster's Kubernetes API", n.GetName()))
+	}
+
+	return nil
+}
+
+func doesNodeExistAsPod(pods []corev1.Pod, n corev1.Node) bool {
```

Comment:
Not sure how often we need this, but there is also second use case, when we want to clean up node. When pod is not in Running phase. this covers situations, when pod crashed or stuck in terminating for some reason.

https://github.com/giantswarm/kvm-operator-node-controller/blob/master/provider/instances.go#L70

Reply:
Is there a test case for this? I ensured all two test cases you mentioned in the original issue are covered. Anyway, this support is super easy to add. Not here, but above in the loop where we already check for the ready condition. Thanks for the hint. I will add this.

Reply:
Done.